zerver/lib/message.py

```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
 
     cache_set_many(items_for_remote_cache)
     return message_ids
+
+
+def visibility_policy_for_participation(
+    sender: UserProfile,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a user
+    participates in a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    if sender.is_bot:
+        return None
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return None
+
+
+def visibility_policy_for_send_message(
+    sender: UserProfile,
+    message: Message,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a message
+    is sent to a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    # We prioritize 'FOLLOW' over 'UNMUTE'.
+    # We need to carefully handle the following two cases:
+    #
+    # 1. When a user starts a topic
+    #    note: starting a topic is INITIATION as well as PARTICIPATION
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |           ON_INITIATION              |     FOLLOWED
+    #         ON_INITIATION                |          ON_PARTICIPATION            |     FOLLOWED
+    #         ON_INITIATION                |            ON_INITIATION             |     FOLLOWED
+    #
+    # 2. When both the policies have the same value 'ON_PARTICIPATION'.
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |          ON_PARTICIPATION            |     FOLLOWED
+    visibility_policy = None
+
+    if sender.is_bot:
+        return visibility_policy
+
+    visibility_policy = visibility_policy_for_participation(sender)
+
+    if visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    accessible_messages_in_topic = UserMessage.objects.filter(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    )
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
+        and accessible_messages_in_topic.count() == 1
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
```

Comment:
I think it might be worth adding an `if visibility_policy == UserTopic.VisibilityPolicy.UNMUTED: return ...` just after this -- the query for how many messages there are can be expensive, and so we should avoid doing it when not necessary.

Reply:
We're now using `.exclude(id=message.id).exists()`; 

It will be called only when:
* automatically_follow_topics_policy == AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
* automatically_unmute_topics_policy == AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION

overall one db query when either or both of the policy is `ON_INITIATION`.

and it is necessary to check if it is the topic start when the policy is `ON_INITIATION`

So, we can avoid adding `if visibility_policy == UserTopic.VisibilityPolicy.UNMUTED: return ...`

Reply:
I think we also want to avoid doing that calculation if we know that the user already has the state that this would trigger. Otherwise, if you send a message to a topic 10 times in a row, we can potentially be doing this query for 9/10 of them for no reason.

Reply:
I think things would be much easier in future when we'll have a `Topic` or `StreamTopic` table. We can simply use a `is_new_topic = True` variable parallely while creating a new `topic` object in the send message codepath; that would lead to avoiding this unnecessart repetitive query.

For now, I'm not sure what would be the ideal place to store the info "is_new_topic = False" after a user finishes sending the first message and reuse during next send message operation.

>  if we know that the user already has the state that this would trigger.

The problem would be for example: If a user has
* follow_policy = on_initiation
* unmute_policy = never

Now if the user sends a message to an existing topic & they have the no visibility policy set; the db query will occur even if we apply the check "if we know that the user already has the state that this would trigger. "

Reply:
@timabbott 
ICYMI: we have one more question to discuss!

Reply:
Regarding a topic table I think we won't be able to do an `is_new_topic` check that way for private streams with protected history (since we can't leak to the user whether there are previous messages); but those are rare, and in shared history streams, it will be a nice optimization. Just wanted to mention that's a corner case that a topics table doesn't make any less ugly.

> Now if the user sends a message to an existing topic & they have the no visibility policy set; the db query will occur even if we apply the check "if we know that the user already has the state that this would trigger. "

I think the fix for this is that we need to pass `current_visibility_policy` to this function, and we can skip fetching existing messages if the user's current_visibility_policy for the topic is already at least as visible as what the policy would change it to.

Reply:
A better fact that we can use to solve this is:

"If a topic has a visibility policy set, it can't be the case of initiation. We return early, thus saving a DB query."

We now pass `current_visibility_policy` to this function and return None if it can't increase visibility.

`should_change_visibility_policy` is now only used in reactions and submessage codepath.

Reply:
OK, this is pretty good! This should probably be a follow-up for sanity's sake, but we might want to make sure we have tests for the query counts for adding a reaction / etc., maybe in functions like ` test_automatically_unmute_topic_on_participation_add_reaction`?